CFE-3046: Only add unique IPv4 interfaces to sys.interfaces

oleorhagen opened the pull request:
Previously, when an interface had multiple IPv4 addresses set, the parsing of
these addresses would add the interface to the sys.interfaces and sys.hardware
variables multiple times, resulting in the interface showing up as a duplicate
in the list.

This fix makes sure that an interface is only added once for every interface
type, regardless of how many IPv4 addresses the interface contains.

Changelog: Title

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

oleorhagen commented:
Sort of fugly fix, but I did not want to change the rest of the loop logic up (something tells me it could have unintended consequences), so I decided to feed the beast some more.

olehermanse commented:
@oleorhagen Looks good, can you paste some terminal output of before and after, showing it working?

oleorhagen commented:
# Recreation

## No fix applied

root@dev masterfiles $ ip a s
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 02:b7:5c:02:92:a2 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s3
       valid_lft 84362sec preferred_lft 84362sec
    inet 10.0.2.16/24 scope global secondary enp0s3
       valid_lft forever preferred_lft forever
    inet6 fe80::b7:5cff:fe02:92a2/64 scope link 
       valid_lft forever preferred_lft forever
3: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 08:00:27:1b:47:6b brd ff:ff:ff:ff:ff:ff
    inet 192.168.100.10/24 brd 192.168.100.255 scope global enp0s8
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe1b:476b/64 scope link 
       valid_lft forever preferred_lft forever
4: enp0s9: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 08:00:27:9e:25:24 brd ff:ff:ff:ff:ff:ff
    inet6 fde4:8dba:82e1::c4/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe9e:2524/64 scope link 
       valid_lft forever preferred_lft forever

root@dev masterfiles $ cf-agent -B 192.168.100.10
R: Bootstrapping from host '192.168.100.10' via built-in policy '/var/cfengine/inputs/failsafe.cf'
R: This host assumes the role of policy server
R: Updated local policy from policy server
R: Triggered an initial run of the policy
R: Restarted systemd unit cfengine3
  notice: Bootstrap to '192.168.100.10' completed successfully!

root@dev masterfiles $ cf-promises --show-vars=sys.interfaces$
Variable name                            Variable value                                               Meta tags                               
default:sys.interfaces                    {"enp0s3","enp0s3","enp0s8","enp0s9"}                       inventory,source=agent,attribute_name=Interfaces

root@dev masterfiles $ cf-promises --show-vars=sys.hardware
Variable name                            Variable value                                               Meta tags                               
default:sys.hardware_addresses            {"02:b7:5c:02:92:a2","02:b7:5c:02:92:a2","08:00:27:1b:47:6b","08:00:27:9e:25:24"} inventory,source=agent,attribute_name=MAC addresses
default:sys.hardware_flags                {"up loopback running","up broadcast running multicast","up broadcast running multicast","up broadcast running multicast"} source=agent                            
default:sys.hardware_mac[enp0s3]         02:b7:5c:02:92:a2                                            source=agent                            
default:sys.hardware_mac[enp0s8]         08:00:27:1b:47:6b                                            source=agent                            
default:sys.hardware_mac[enp0s9]         08:00:27:9e:25:24                                            source=agent,derived-from=ifconfig      

## With fix applied

root@dev core $ cf-agent -B 192.168.100.10
R: Bootstrapping from host '192.168.100.10' via built-in policy '/var/cfengine/inputs/failsafe.cf'
R: This host assumes the role of policy server
R: Updated local policy from policy server
R: Triggered an initial run of the policy
  notice: Bootstrap to '192.168.100.10' completed successfully!
root@dev core $ cf-promises --show-vars=sys.interfaces$
Variable name                            Variable value                                               Meta tags                               
default:sys.interfaces                    {"enp0s3","enp0s8","enp0s9"}                                inventory,source=agent,attribute_name=Interfaces
root@dev core $ cf-promises --show-vars=sys.hardware
Variable name                            Variable value                                               Meta tags                               
default:sys.hardware_addresses            {"02:b7:5c:02:92:a2","08:00:27:1b:47:6b","08:00:27:9e:25:24"} inventory,source=agent,attribute_name=MAC addresses
default:sys.hardware_flags                {"up loopback running","up broadcast running multicast","up broadcast running multicast","up broadcast running multicast"} source=agent                            
default:sys.hardware_mac[enp0s3]         02:b7:5c:02:92:a2                                            source=agent                            
default:sys.hardware_mac[enp0s8]         08:00:27:1b:47:6b                                            source=agent                            
default:sys.hardware_mac[enp0s9]         08:00:27:9e:25:24                                            source=agent,derived-from=ifconfig      

## Notable differences:

| variable/step | sys.interfaces | sys.hardware_addresses |
| :---  | :--- | :--- |
| Pre | {"enp0s3","enp0s3","enp0s8","enp0s9"} | {"02:b7:5c:02:92:a2","02:b7:5c:02:92:a2","08:00:27:1b:47:6b","08:00:27:9e:25:24"} |
| Post | {"enp0s3","enp0s8","enp0s9"} | {"02:b7:5c:02:92:a2","08:00:27:1b:47:6b","08:00:27:9e:25:24"} |

olehermanse commented:
@cf-bottom test this in jenkins, please.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=4137)](https://ci.cfengine.com//job/pr-pipeline/4137/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4137/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4137/

nickanderson commented:
![carlton](https://user-images.githubusercontent.com/202896/68872989-83a1d800-06c4-11ea-8813-92826fcd3899.gif)

olehermanse commented:
Cherry-pick: https://github.com/cfengine/core/pull/3946
